atc0005/check-cert
# Add configurable apply/ignore validation support

## OVERVIEW

Prior versions of the `check_cert` plugin and `lscert` CLI tool
applied validation checks conditionally, noting validation check
results only when performed/applied. While logical, this didn't
consistently answer the question a sysadmin might have for what checks
*were* performed and just did not trigger an error.

This commit (and associated PR) reworks validation checks handling in
order to:

- make it clearer which validation checks are performed
- make it clearer which validation checks resulted in the final plugin
  state
- give sysadmins more control over which validation checks should be
  considered when determining final plugin state
- allow for easier extension of validation check logic in the future
  without increasing the risk of breaking existing monitoring
  configurations

In contrast to earlier versions, the `check_cert` plugin and `lscert`
CLI tool perform *all* validation checks, but will flag the results as
ignored when specific requirements are not met. Support has been added
to the `check_cert` plugin to explicitly *apply* or *ignore* specific
validation check results.

For example, the SANs list validation check (asserts that specified
SANs entries are present) is performed by default, but if the sysadmin
does not specify SANs entries to use for validation, only minimal
validation can be applied. This validation check result is flagged as
ignored.

For the `check_cert` plugin:

- the check result will not be used when determining the final plugin
  state
- this decision will be noted in the final plugin output

For the `lscert` CLI tool:

- this ignored status is listed as "OK" in the check results summary

No changes in behavior were made for the `certsum` CLI tool. Future
work may extend the tool to flag "problem" certificates using a range
of validation checks in addition to the current certificate expiration
validation.

## CHANGES

The set of changes included in this commit/PR is extensive. While the
primary focus was refactoring the validation check logic, many smaller
changes were made to polish rough edges and correct long-standing
issues. Some of those changes are noted below.

See also the recent work on the atc0005/go-nagios package. Many
changes were made in tandem between the v0.9.x release of that package
and the work in this project reflected in this commit/PR. In
particular, since it was known that the validation check results work
would increase plugin output significant effort was put into refining
existing output and dropping details found not to be as useful.

- heavily rework validation logic (as noted in the overview) to allow
  all validation checks to execute regardless of an earlier validation
  check's result
  - for example, if a hostname validation check result indicates
    failure, proceed with checking certificate expiration also and
    report together vs immediately declaring failure and ending plugin
    execution
- add new certs.CertChainValidationResult interface
  - this forms the backbone of the new validation check handling and
    is intended to allow extending validation check behavior without
    breaking existing monitoring configurations in the future
- drop Thresholds section from `check_cert` plugin
  - follows changes recently made to the `lscert` CLI tool
- change default LongServiceOutput label to help make clear that the
  output is intended as a report
- change errors section label to match LongServiceOutput label
- clarify certificate source (file vs service)
- replace a lot of no longer used code in the internal/certs package
  related to the certs.ChainStatus type
- add tests for new validation check apply/ignore logic, config
  validation logic
- increase debug message detail for each validation check performed
- replace logging related globals, init function, config validation in
  internal/config package
- replace hard-coded flag names with constants and use those contacts
  from applicable flag help text, config validation output
- extend flag help text to dynamically provide a list of supported
  values for flags using a fixed set of supported configuration
  keywords (e.g., the new apply/ignore logic keywords for specific
  validation checks)
- deprecate `disable-hostname-verification-if-empty-sans` flag
- documentation refresh
  - highlight features of each tool provided by this project,
    changes from this work, SNI support/behavior applied by the use of
    the `server` and `dns-name` flags
  - add Quick Start guide to install latest version of each tool using
    `go install` along with the `latest` tag keyword

There are likely other changes as well, but these are the more
significant of the bunch.

## REFERENCES

- fixes GH-211
- fixes GH-315
- refs GH-356
